File: min-cost-climbing-stairs/min-cost-climbing-stairs.cpp
class Solution {
public:
    int dp[1005];
    int minCostClimbingStairs(vector<int>& cost) {
        for(int i = 0; i < 1005; i++)
            dp[i] = -1;
        
        return min(solve(cost, 0), solve(cost, 1));
    }
    
    int solve(vector<int>&cost, int pos)
    {
        if(pos >= cost.size()){
            return 0;
        }
        
        if(dp[pos] != -1) {
            return dp[pos];
        }
        
        int q1 = cost[pos] + solve(cost, pos + 1);
        int q2 = cost[pos] + solve(cost, pos + 2);
        
        return dp[pos] = min(q1, q2);
    }
};

File: min-cost-climbing-stairs/min-cost-climbing-stairs.py
class Solution:
    def __init__(self):
        self.dp = [-1 for i in range(1010)]
        
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        self.cost = cost
        return min(self.solve(0), self.solve(1))
    
    def solve(self, pos) -> int:
        if pos >= len(self.cost):
            return 0
        
        if self.dp[pos] != -1:
            return self.dp[pos]
        
        self.dp[pos] = self.cost[pos] + min(self.solve(pos + 1), self.solve(pos + 2))
        return self.dp[pos]
        